### DISCUSSION TITLE
Does Rsbuild needs to support using Webpack?

### DISCUSSION THREAD

**Discussion by chenjiahan** [web-infra-dev/rsbuild#239] [General]
## The problem

In the Modern.js framework, users can choose to use Webpack or Rspack as the bundler, so Modern.js Builder needs to support both of them.

When developing the new Rsbuild, we found that it is difficult to compat Webpack, especially when we need to support new frameworks like Vue and Svelte, and support both Babel and SWC as the transformers.

Imagine these combinations, they will bring a high level of complexity and mess the code:

- Webpack + SWC + React
- Webpack + Babel + React
- Webpack + SWC + Vue
- Webpack + Babel + Vue
- Rspack + SWC (builtin) + React
- Rspack + Babel + React
- Rspack + SWC (builtin) + Vue
- Rspack + Babel + Vue
- more...

## New design

To avoid this situation, we are considering dropping support for Webpack in Rsbuild. This means that Rsbuild users can only use one bundler —— Rspack.

The combinations can be reduced to:

- Rspack + SWC (builtin) + React
- Rspack + SWC (builtin) + Vue
- Rspack + Babel + Vue (Vue JSX relies on Babel)

And for Modern.js (PIA) users, Webpack & Babel is still available.  We will merge the code that is compatible with Webpack + Babel + React as a private package for Modern.js (PIA) users, which will not be exposed to Rsbuild users.

## The benefits

- More in line with a long-term correct architecture.
- Make it easier for Rsbuild to support more frameworks and features.
- Reduce Rsbuild's dependencies, making it more lightweight.
- Maintain compatibility for Modern.js (PIA) users.

**Comment by hardfist** [comment]
strongly agree

**Comment by chenjiahan** [comment]
The architecture after adjustment:

![unibuilder](https://github.com/web-infra-dev/rsbuild/assets/7237365/1308e184-f4fc-4722-8390-b2bc343b93ca)

The `@rsbuild/unibuilder` package will provide the same capability as `@modern-js/builder` + `@modern-js/builder-webpack-provider` + `@modern-js/rspack-provider`.
